README.md
<p align="center"><a href="https://github.com/crazy-max/docker-jetbrains-license-server" target="_blank"><img height="128" src="https://raw.githubusercontent.com/crazy-max/docker-jetbrains-license-server/master/.github/docker-jetbrains-license-server.jpg"></a></p>

<p align="center">
  <a href="https://hub.docker.com/r/crazymax/jetbrains-license-server/tags?page=1&ordering=last_updated"><img src="https://img.shields.io/github/v/tag/crazy-max/docker-jetbrains-license-server?label=version&style=flat-square" alt="Latest Version"></a>
  <a href="https://github.com/crazy-max/docker-jetbrains-license-server/actions?workflow=build"><img src="https://img.shields.io/github/actions/workflow/status/crazy-max/docker-jetbrains-license-server/build.yml?branch=master&?label=build&logo=github&style=flat-square" alt="Build Status"></a>
  <a href="https://hub.docker.com/r/crazymax/jetbrains-license-server/"><img src="https://img.shields.io/docker/stars/crazymax/jetbrains-license-server.svg?style=flat-square&logo=docker" alt="Docker Stars"></a>
  <a href="https://hub.docker.com/r/crazymax/jetbrains-license-server/"><img src="https://img.shields.io/docker/pulls/crazymax/jetbrains-license-server.svg?style=flat-square&logo=docker" alt="Docker Pulls"></a>
  <br /><a href="https://github.com/sponsors/crazy-max"><img src="https://img.shields.io/badge/sponsor-crazy--max-181717.svg?logo=github&style=flat-square" alt="Become a sponsor"></a>
  <a href="https://www.paypal.me/crazyws"><img src="https://img.shields.io/badge/donate-paypal-00457c.svg?logo=paypal&style=flat-square" alt="Donate Paypal"></a>
</p>

## About

[JetBrains License Server](https://www.jetbrains.com/help/license_server/getting_started.html)
Docker image.

> **Note**
> 
> Want to be notified of new releases? Check out 🔔 [Diun (Docker Image Update Notifier)](https://github.com/crazy-max/diun)
> project!

___

* [Features](#features)
* [Build locally](#build-locally)
* [Image](#image)
* [Environment variables](#environment-variables)
* [Volumes](#volumes)
* [Ports](#ports)
* [Usage](#usage)
  * [Docker Compose](#docker-compose)
  * [Command line](#command-line)
* [Upgrade](#upgrade)
* [Notes](#notes)
  * [Error 403 Passed value of header "Host" is not allowed](#error-403-passed-value-of-header-host-is-not-allowed)
* [Contributing](#contributing)
* [License](#license)

## Features

* Run as non-root user
* Multi-platform image
* License server completely customizable via environment variables
* Registration data and configuration in a single directory
* [msmtpd SMTP relay](https://github.com/crazy-max/docker-msmtpd) image to send emails
* [Traefik](https://github.com/containous/traefik-library-image) as reverse proxy and creation/renewal of Let's Encrypt certificates (see [this template](examples/traefik))

## Build locally

```shell
git clone https://github.com/crazy-max/docker-jetbrains-license-server.git
cd docker-jetbrains-license-server

# Build image and output to docker (default)
docker buildx bake

# Build multi-platform image
docker buildx bake image-all
```

## Image

| Registry                                                                                                            | Image                                        |
|---------------------------------------------------------------------------------------------------------------------|----------------------------------------------|
| [Docker Hub](https://hub.docker.com/r/crazymax/jetbrains-license-server/)                                           | `crazymax/jetbrains-license-server`          |
| [GitHub Container Registry](https://github.com/users/crazy-max/packages/container/package/jetbrains-license-server) | `ghcr.io/crazy-max/jetbrains-license-server` |

## Environment variables

* `TZ`: The timezone assigned to the container (default `UTC`)
* `PUID`: JLS UID (default `1000`)
* `PGID`: JLS GID (default `1000`)
* `JLS_VIRTUAL_HOSTS`: [Virtual hosts](https://www.jetbrains.com/help/license_server/setting_host_and_port.html#d1010e63) where license server will be available (comma delimited for several hosts)
* `JLS_CONTEXT`:  [Context path](https://www.jetbrains.com/help/license_server/setting_host_and_port.html#d1010e63) used by the license server (default `/`)
* `JLS_PROXY_TYPE`: Type of [proxy](https://www.jetbrains.com/help/license_server/configuring_proxy_settings.html) to use. Can be `http` or `https` (default `https`)
* `JLS_PROXY_HOST`: The host name of your proxy server
* `JLS_PROXY_PORT`: The port number that the proxy server listens to
* `JLS_PROXY_USER`: Username to connect to the proxy server (no auth if empty)
* `JLS_PROXY_PASSWORD`: Password to connect to the proxy server (no auth if empty)
* `JLS_ACCESS_CONFIG`: JSON file to configure [user restrictions](https://www.jetbrains.com/help/license_server/configuring_user_restrictions.html) (default `/data/access-config.json`)
* `JLS_STATS_RECIPIENTS`: [Reports recipients](https://www.jetbrains.com/help/license_server/detailed_server_usage_statistics.html#d461e40) email addresses for stats (comma delimited)
* `JLS_REPORT_OUT_OF_LICENSE`: [Warn about lack of licenses](https://www.jetbrains.com/help/license_server/detailed_server_usage_statistics.html#d461e40) every hour following the percentage threshold (default `0`)
* `JLS_SMTP_SERVER`: SMTP server host to use for sending [stats](https://www.jetbrains.com/help/license_server/detailed_server_usage_statistics.html) (stats disabled if empty)
* `JLS_SMTP_PORT`: SMTP server port (default `25`)
* `JLS_SMTP_USERNAME`: SMTP username (auth disabled if empty)
* `JLS_SMTP_PASSWORD`: SMTP password (auth disabled if empty)
* `JLS_STATS_FROM`: [From address](https://www.jetbrains.com/help/license_server/detailed_server_usage_statistics.html#d461e40) for stats emails
* `JLS_STATS_TOKEN`: Enables an auth token for the [stats API](https://www.jetbrains.com/help/license_server/detailed_server_usage_statistics.html#d461e312) at `/reportApi` (HTTP POST)
* `JLS_SERVICE_LOGLEVEL`: Logging of IDE requests and responses (default `warn`)
* `JLS_REPORTING_LOGLEVEL`: Logging of actions on a server: tickets obtaining and revoking with user and license information (default `warn`) 
* `JLS_TICKETS_LOGLEVEL`: Logging of actions with tickets. For example, manual ticket revoking (default `warn`)

## Volumes

* `/data`: Contains [registration data](https://www.jetbrains.com/help/license_server/migrate.html) and configuration

> **Warning**
>
> Note that the volumes should be owned by the user/group with the specified
> `PUID` and `PGID`. If you don't give the volume correct permissions, the
> container may not start.

## Ports

* `8000`: Jetbrains License Server HTTP port

## Usage

### Docker Compose

Docker compose is the recommended way to run this image. Copy the content of
folder [examples/compose](examples/compose) in `/var/jls/` on your host for
example. Edit the compose and env files with your preferences and run the
following commands:

```bash
docker compose up -d
docker compose logs -f
```

### Command line

You can also use the following minimal command:

```bash
$ docker run -d -p 8000:8000 --name jetbrains_license_server \
  -e TZ="Europe/Paris" \
  -e JLS_VIRTUAL_HOSTS=jls.example.com \
  -v $(pwd)/data:/data \
  crazymax/jetbrains-license-server:latest
```

## Upgrade

Recreate the container whenever I push an update:

```bash
docker compose pull
docker compose up -d
```

## Notes

If you have any trouble using the license server, check the official
[Troubleshooting page](https://www.jetbrains.com/help/license_server/troubleshooting.html)
of Jetbrains.

### Error 403 Passed value of header "Host" is not allowed

If you've got the following message :

```
Passed value of header "Host" is not allowed. Please contact your license server administrator.
```

That's because the license server is running behind a reverse proxy. Please
configure virtual hosts using the `JLS_VIRTUAL_HOSTS` variable.

## Contributing

Want to contribute? Awesome! The most basic way to show your support is to star
the project, or to raise issues. You can also support this project by [**becoming a sponsor on GitHub**](https://github.com/sponsors/crazy-max)
or by making a [PayPal donation](https://www.paypal.me/crazyws) to ensure this
journey continues indefinitely!

Thanks again for your support, it is much appreciated! :pray:

## License

MIT. See `LICENSE` for more details.


CHANGELOG.md
# Changelog

## 39687-r0 (2023/12/30)

* JetBrains License Server 39687 (#91)
* Alpine Linux 3.19 (#92)

## 33965-r1 (2023/08/30)

* Alpine Linux 3.18 (#82)

## 33965-r0 (2022/12/29)

* JetBrains License Server 33965 (#75)
* Alpine Linux 3.17 (#74)

## 32800-r0 (2022/07/20)

* JetBrains License Server 32800 (#70)
* Alpine Linux 3.16 (#71)

## 31550-r0 (2022/04/10)

* JetBrains License Server 31550 (#64)

## 30803-r0 (2022/02/07)

* JetBrains License Server 30803 (#61)
* OpenJDK 11 required (drops support for `arm/v6` and `arm/v7`)

## 30333-r0 (2021/12/22)

* JetBrains License Server 30333 (#60)

## 30241-r0 (2021/12/17)

* JetBrains License Server 30241 (#59)

## 30211-r0 (2021/12/12)

* JetBrains License Server 30211 (#57)
* Alpine Linux 3.15 (#58)

## 29591-r0 (2021/10/26)

* JetBrains License Server 29591 (#56)

## 28662-r1 (2021/09/18)

* Alpine Linux 3.14 (#55)
* Fix JRE (#54)

## 28662-r0 (2021/08/15)

* JetBrains License Server 28662

## 28150-r0 (2021/07/14)

* JetBrains License Server 28150

## 27608-r0 (2021/06/29)

* JetBrains License Server 27608

## 26498-r0 (2021/03/17)

* JetBrains License Server 26498

## 26301-r3 (2021/03/04)

* Renamed `yasu` (more info https://github.com/crazy-max/yasu#yet-another)

## 26301-r2 (2021/03/03)

* Switch to `gosu`

## 26301-r1 (2021/02/21)

* Fix permissions (#44)

## 26301-r0 (2021/02/11)

* JetBrains License Server 26301

## 25980-r0 (2021/01/18)

* JetBrains License Server 25980
* OpenJDK JRE 15
* Switch to buildx bake

## 24694-RC1 (2020/10/08)

* JetBrains License Server 24694
* Add `JLS_SERVICE_LOGLEVEL`, `JLS_REPORTING_LOGLEVEL` and `JLS_TICKETS_LOGLEVEL` env vars

## 24086-RC1 (2020/08/17)

* JetBrains License Server 24086

## 23527-RC1 (2020/06/19)

* JetBrains License Server 23527

## 22218-RC1 (2020/02/04)

* JetBrains License Server 22218
* AdoptOpenJDK JRE 13

## 21137-RC5 (2019/12/07)

* Fix timezone

## 21137-RC4 (2019/11/28)

* Proxy type defaults to `https`
* More unset

## 21137-RC3 (2019/11/28)

* Add env vars to set proxy settings (#21)
* Unset sensitive vars

## 21137-RC2 (2019/11/17)

* Allow to set custom `PUID`/`PGID`

## 21137-RC1 (2019/10/24)

* JetBrains License Server 21137

## 20308-RC4 (2019/10/10)

* Optimize layers

## 20308-RC3 (2019/10/10)

* Multi-platform Docker image
* Switch to GitHub Actions
* :warning: Stop publishing Docker image on Quay
* :warning: Run as non-root user
* Set timezone through tzdata

> :warning: **UPGRADE NOTES**
> As the Docker container now runs as a non-root user, you have to first stop the container and change permissions to `data` volume:
> ```
> docker compose stop
> chown -R 1000:1000 data/
> docker compose pull
> docker compose up -d
> ```

## 20308-RC2 (2019/08/04)

* Add healthcheck

## 20308-RC1 (2019/07/30)

* JetBrains License Server 20308

## 20267-RC2 (2019/07/22)

* OpenJDK JRE 12
* Alpine Linux 3.10

## 20267-RC1 (2019/07/19)

* JetBrains License Server 20267

## 19488-RC1 (2019/04/02)

* JetBrains License Server 19488

## 19340-RC1 (2019/03/15)

* JetBrains License Server 19340

## 18692-RC2 (2019/01/22)

* Bind to unprivileged port : `8000`

## 18692-RC1 (2018/12/25)

* JetBrains License Server 18692

## 17955-RC1 (2018/09/26)

* JetBrains License Server 17955

## 17768-RC1 (2018/09/05)

* JetBrains License Server 17768

## 17437-RC1 (2018/07/30)

* JetBrains License Server 17437

## 17211-RC1 (2018/06/29)

* JetBrains License Server 17211

## 17043-RC1 (2018/05/31)

* JetBrains License Server 17043

## 16429-RC2 (2018/04/20)

* JetBrains accidentally published 16743. Revert to 16429 (Issue #1)

## 16743-RC1 (2018/04/20)

* JetBrains License Server 16743
* Replace Nginx + Let's Encrypt with Traefik (see docker-compose)

## 16429-RC1 (2018/03/20)

* JetBrains License Server 16429

## 15802-RC2 (2018/02/16)

* Add `JLS_CONTEXT` and `JLS_REPORT_OUT_OF_LICENSE` env vars
* Timezone was not setted
* No need of Nginx and Supervisor
* Error while saving stats
* Remove build dependencies

## 15802-RC1 (2018/02/01)

* Initial version based on JetBrains License Server 15802


build.sh
#!/usr/bin/env bash
set -e

PROJECT=jetbrains-license-server
BUILD_DATE=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
BUILD_TAG=docker_build
BUILD_WORKINGDIR=${BUILD_WORKINGDIR:-.}
DOCKERFILE=${DOCKERFILE:-Dockerfile}
VCS_REF=${TRAVIS_COMMIT::7}
RUNNING_TIMEOUT=120
RUNNING_LOG_CHECK="is listening on"

PUSH_LATEST=${PUSH_LATEST:-true}
DOCKER_USERNAME=${DOCKER_USERNAME:-crazymax}
DOCKER_LOGIN=${DOCKER_LOGIN:-crazymax}
DOCKER_REPONAME=${DOCKER_REPONAME:-jetbrains-license-server}
QUAY_USERNAME=${QUAY_USERNAME:-crazymax}
QUAY_LOGIN=${QUAY_LOGIN:-crazymax}
QUAY_REPONAME=${QUAY_REPONAME:-jetbrains-license-server}

# Check local or travis
BRANCH=${TRAVIS_BRANCH:-local}
if [[ ${TRAVIS_PULL_REQUEST} == "true" ]]; then
  BRANCH=${TRAVIS_PULL_REQUEST_BRANCH}
fi
DOCKER_TAG=${BRANCH:-local}
if [[ "$BRANCH" == "master" ]]; then
  DOCKER_TAG=latest
elif [[ "$BRANCH" == "local" ]]; then
  BUILD_DATE=
  VERSION=local
fi

echo "PROJECT=${PROJECT}"
echo "VERSION=${VERSION}"
echo "BUILD_DATE=${BUILD_DATE}"
echo "BUILD_TAG=${BUILD_TAG}"
echo "BUILD_WORKINGDIR=${BUILD_WORKINGDIR}"
echo "DOCKERFILE=${DOCKERFILE}"
echo "VCS_REF=${VCS_REF}"
echo "PUSH_LATEST=${PUSH_LATEST}"
echo "DOCKER_LOGIN=${DOCKER_LOGIN}"
echo "DOCKER_USERNAME=${DOCKER_USERNAME}"
echo "DOCKER_REPONAME=${DOCKER_REPONAME}"
echo "QUAY_LOGIN=${QUAY_LOGIN}"
echo "QUAY_USERNAME=${QUAY_USERNAME}"
echo "QUAY_REPONAME=${QUAY_REPONAME}"
echo "TRAVIS_BRANCH=${TRAVIS_BRANCH}"
echo "TRAVIS_PULL_REQUEST=${TRAVIS_PULL_REQUEST}"
echo "BRANCH=${BRANCH}"
echo "DOCKER_TAG=${DOCKER_TAG}"
echo

# Build
echo "### Build"
docker build \
  --build-arg BUILD_DATE=${BUILD_DATE} \
  --build-arg VCS_REF=${VCS_REF} \
  --build-arg VERSION=${VERSION} \
  -t ${BUILD_TAG} -f ${DOCKERFILE} ${BUILD_WORKINGDIR}
echo

echo "### Test"
docker rm -f ${PROJECT} > /dev/null 2>&1 || true
docker run -d -p 8000:8000 \
  -e "JLS_VIRTUAL_HOSTS=docker.test" \
  --name ${PROJECT} ${BUILD_TAG}
echo

echo "### Waiting for ${PROJECT} to be up..."
TIMEOUT=$((SECONDS + RUNNING_TIMEOUT))
while read LOGLINE; do
  echo ${LOGLINE}
  if [[ ${LOGLINE} == *"${RUNNING_LOG_CHECK}"* ]]; then
    echo "Container up!"
    break
  fi
  if [[ $SECONDS -gt ${TIMEOUT} ]]; then
    >&2 echo "ERROR: Failed to run ${PROJECT} container"
    docker rm -f ${PROJECT} > /dev/null 2>&1 || true
    exit 1
  fi
done < <(docker logs -f ${PROJECT} 2>&1)
echo

CONTAINER_STATUS=$(docker container inspect --format "{{.State.Status}}" ${PROJECT})
if [[ ${CONTAINER_STATUS} != "running" ]]; then
  >&2 echo "ERROR: Container ${PROJECT} returned status '$CONTAINER_STATUS'"
  docker rm -f ${PROJECT} > /dev/null 2>&1 || true
  exit 1
fi
docker rm -f ${PROJECT} > /dev/null 2>&1 || true

if [ "${VERSION}" == "local" -o "${TRAVIS_PULL_REQUEST}" == "true" ]; then
  echo "INFO: This is a PR or a local build, skipping push..."
  exit 0
fi
if [[ ! -z ${DOCKER_PASSWORD} ]]; then
  echo "### Push to Docker Hub..."
  echo "$DOCKER_PASSWORD" | docker login --username "$DOCKER_LOGIN" --password-stdin > /dev/null 2>&1
  if [ "${DOCKER_TAG}" == "latest" -a "${PUSH_LATEST}" == "true" ]; then
    docker tag ${BUILD_TAG} ${DOCKER_USERNAME}/${DOCKER_REPONAME}:${DOCKER_TAG}
  fi
  if [[ "${VERSION}" != "latest" ]]; then
    docker tag ${BUILD_TAG} ${DOCKER_USERNAME}/${DOCKER_REPONAME}:${VERSION}
  fi
  docker push ${DOCKER_USERNAME}/${DOCKER_REPONAME}
  echo
fi
if [[ ! -z ${QUAY_PASSWORD} ]]; then
  echo "### Push to Quay..."
  echo "$QUAY_PASSWORD" | docker login quay.io --username "$QUAY_LOGIN" --password-stdin > /dev/null 2>&1
  if [ "${DOCKER_TAG}" == "latest" -a "${PUSH_LATEST}" == "true" ]; then
    docker tag ${BUILD_TAG} quay.io/${QUAY_USERNAME}/${QUAY_REPONAME}:${DOCKER_TAG}
  fi
  if [[ "${VERSION}" != "latest" ]]; then
    docker tag ${BUILD_TAG} quay.io/${QUAY_USERNAME}/${QUAY_REPONAME}:${VERSION}
  fi
  docker push quay.io/${QUAY_USERNAME}/${QUAY_REPONAME}
  echo
fi


entrypoint.sh
#!/bin/sh

# SSH server inside the container
#service ssh start

# Nginx for fixing HTTP headers modified by Azure's App Service proxy
#service nginx start
rc-service nginx start

TZ=${TZ:-UTC}

JLS_PATH="/opt/jetbrains-license-server"
JLS_LISTEN_ADDRESS="0.0.0.0"
JLS_PORT=8000
JLS_CONTEXT=${JLS_CONTEXT:-/}
JLS_ACCESS_CONFIG=${JLS_ACCESS_CONFIG:-/home/access-config.json}
JLS_PROXY_TYPE=${JLS_PROXY_TYPE:-https}
JLS_SERVICE_LOGLEVEL=${JLS_SERVICE_LOGLEVEL:-warn}
JLS_REPORTING_LOGLEVEL=${JLS_REPORTING_LOGLEVEL:-warn}
JLS_TICKETS_LOGLEVEL=${JLS_TICKETS_LOGLEVEL:-warn}

if [ -n "${PGID}" ] && [ "${PGID}" != "$(id -g jls)" ]; then
  echo "Switching to PGID ${PGID}..."
  sed -i -e "s/^jls:\([^:]*\):[0-9]*/jls:\1:${PGID}/" /etc/group
  sed -i -e "s/^jls:\([^:]*\):\([0-9]*\):[0-9]*/jls:\1:\2:${PGID}/" /etc/passwd
fi
if [ -n "${PUID}" ] && [ "${PUID}" != "$(id -u jls)" ]; then
  echo "Switching to PUID ${PUID}..."
  sed -i -e "s/^jls:\([^:]*\):[0-9]*:\([0-9]*\)/jls:\1:${PUID}:\2/" /etc/passwd
fi
if [ -n "${PUID}" ] || [ -n "${PGID}" ]; then
  chown -R jls:jls /home "$JLS_PATH"
fi

# Timezone
echo "Setting timezone to ${TZ}..."
ln -snf /usr/share/zoneinfo/${TZ} /etc/localtime
echo ${TZ} > /etc/timezone

# Init
echo "Initializing files and folders..."
yasu jls:jls mkdir -p /home/registration
yasu jls:jls touch "/home/access-config.json"

# https://www.jetbrains.com/help/license_server/setting_host_and_port.html
echo "Configuring Jetbrains License Server..."
yasu jls:jls license-server configure --listen ${JLS_LISTEN_ADDRESS} --port ${JLS_PORT} --context ${JLS_CONTEXT}
yasu jls:jls license-server configure --jb.license-server.base-dir=/home/registration

# https://www.jetbrains.com/help/license_server/configuring_secure_connection.html
if [ ! -z "$JLS_REVERSE_PROXY" ] ; then
  echo "Configuring reverse proxy..."
  license-server configure --listen-port 80 --base-url http://${JLS_REVERSE_PROXY}:443
fi

# https://www.jetbrains.com/help/license_server/setting_host_and_port.html
if [ -n "$JLS_VIRTUAL_HOSTS" ]; then
  echo "Following virtual hosts will be used:"
  for JLS_VIRTUAL_HOST in $(echo ${JLS_VIRTUAL_HOSTS} | tr "," "\n"); do
    echo "-> ${JLS_VIRTUAL_HOST}"
  done
  yasu jls:jls license-server configure --jetty.virtualHosts.names=${JLS_VIRTUAL_HOSTS}
fi

# https://www.jetbrains.com/help/license_server/configuring_proxy_settings.html
if [ -n "$JLS_PROXY_HOST" ] && [ -n "$JLS_PROXY_PORT" ]; then
  echo "Setting ${JLS_PROXY_TYPE} proxy to $JLS_PROXY_HOST:$JLS_PROXY_PORT..."
  yasu jls:jls license-server configure \
    -J-D${JLS_PROXY_TYPE}.proxyHost=${JLS_PROXY_HOST} \
    -J-D${JLS_PROXY_TYPE}.proxyPort=${JLS_PROXY_PORT}

  if [ -n "$JLS_PROXY_USER" ] && [ -n "$JLS_PROXY_PASSWORD" ]; then
    echo "Setting ${JLS_PROXY_TYPE} proxy credentials..."
    yasu jls:jls license-server configure \
      -J-D${JLS_PROXY_TYPE}.proxyUser=${JLS_PROXY_USER} \
      -J-D${JLS_PROXY_TYPE}.proxyPassword=${JLS_PROXY_PASSWORD}
  fi
fi
unset JLS_PROXY_USER
unset JLS_PROXY_PASSWORD

# https://www.jetbrains.com/help/license_server/configuring_user_restrictions.html
if [ -s "$JLS_ACCESS_CONFIG" ]; then
  echo "Enabling user restrictions access from $JLS_ACCESS_CONFIG..."
  yasu jls:jls license-server configure --access.config=file:${JLS_ACCESS_CONFIG}
fi

# https://www.jetbrains.com/help/license_server/detailed_server_usage_statistics.html
if [ -n "$JLS_SMTP_SERVER" ] && [ -n "$JLS_STATS_RECIPIENTS" ]; then
  JLS_SMTP_PORT=${JLS_SMTP_PORT:-25}
  echo "Enabling User Reporting via SMTP at $JLS_SMTP_SERVER:$JLS_SMTP_PORT..."
  yasu jls:jls license-server configure --smtp.server ${JLS_SMTP_SERVER} --smtp.server.port ${JLS_SMTP_PORT}

  if [ -n "$JLS_SMTP_USERNAME" ] && [ -n "$JLS_SMTP_PASSWORD" ]; then
    echo "Using SMTP username $JLS_SMTP_USERNAME with password..."
    yasu jls:jls license-server configure --smtp.server.username ${JLS_SMTP_USERNAME}
    yasu jls:jls license-server configure --smtp.server.password ${JLS_SMTP_PASSWORD}
  fi
  unset JLS_SMTP_USERNAME
  unset JLS_SMTP_PASSWORD

  if [ -n "$JLS_STATS_FROM" ]; then
    echo "Setting stats sender to $JLS_STATS_FROM..."
    yasu jls:jls license-server configure --stats.from ${JLS_STATS_FROM}
  fi

  if [ "$JLS_REPORT_OUT_OF_LICENSE" -gt 0 ]; then
    echo "Setting report out of licence to $JLS_REPORT_OUT_OF_LICENSE%..."
    yasu jls:jls license-server configure --reporting.out.of.license.threshold ${JLS_REPORT_OUT_OF_LICENSE}
  fi

  echo "Stats recipients: $JLS_STATS_RECIPIENTS..."
  yasu jls:jls license-server configure --stats.recipients ${JLS_STATS_RECIPIENTS}
fi

# https://www.jetbrains.com/help/license_server/detailed_server_usage_statistics.html
if [ -n "$JLS_STATS_TOKEN" ]; then
  echo "Enabling stats via API at /$JLS_STATS_TOKEN..."
  yasu jls:jls license-server configure --reporting.token ${JLS_STATS_TOKEN}
fi
unset JLS_STATS_TOKEN

# https://www.jetbrains.com/help/license_server/changing_logging_level.html
cat > ${JLS_PATH}/web/WEB-INF/classes/log4j2.xml <<EOL
<?xml version="1.0" encoding="UTF-8"?>
<Configuration status="info" monitorInterval="60">
    <Appenders>
        <Console name="CONSOLE" target="SYSTEM_OUT">
            <PatternLayout pattern="%-5p %c{1}:%L - %m%n"/>
        </Console>
        <Async name="ASYNC" includeLocation="true">
            <AppenderRef ref="CONSOLE"/>
        </Async>
    </Appenders>
    <Loggers>
        <Root level="info">
            <AppenderRef ref="ASYNC"/>
        </Root>
        <Logger name="l_service" level="${JLS_SERVICE_LOGLEVEL}" additivity="false">
            <AppenderRef ref="ASYNC"/>
        </Logger>
        <Logger name="Reporting" level="${JLS_REPORTING_LOGLEVEL}" additivity="false">
            <AppenderRef ref="ASYNC"/>
        </Logger>
        <Logger name="Tickets" level="${JLS_TICKETS_LOGLEVEL}" additivity="false">
            <AppenderRef ref="ASYNC"/>
        </Logger>
    </Loggers>
</Configuration>
EOL

echo "Fixing perms..."
chown -R jls:jls /home "$JLS_PATH"

# https://www.jetbrains.com/help/license_server/configuring_proxy_settings.html
if [ -n "$JLS_PROXY_HOST" ] && [ -n "$JLS_PROXY_PORT" ] ; then
  echo "Enabling HTTP proxy..."
  license-server configure --https.proxyHost ${JLS_PROXY_HOST} --https.proxyPort ${JLS_PROXY_PORT}
fi

exec "$@"
# exec yasu jls:jls "$@"
